### PR TITLE
Optimize CEpgInfoTag. No more copies, just shared pointers.

### PR DESCRIPTION
**Pull request by ksooo**
@opdenkamp This is the EpgInfoTag optimization we discussed in #5629.

### CODE REVIEW

**Comment by Jalle19** [comment]
@opdenkamp would be nice to get this one merged asap so we have time to deal with potential regressions.

@ksooo are you up for making similar changes to some of the other classes? Channels and recordings are everywhere at the moment.

**Comment by opdenkamp** [comment]
yeah will check asap, probably this week end. just checking email spam from the last 2 weeks on github atm.

**Comment by ksooo** [comment]
@Jalle19 : You mean
https://github.com/ksooo/xbmc/tree/recording-optimizations +
https://github.com/ksooo/xbmc/tree/pvrchannel-optimizations +
https://github.com/ksooo/xbmc/tree/pvrtimerinfotag-optimizations ? 

All ptached in my private kodi builds for quite some time now. No issues so far. ;-) Will issue PRs step by step, one after the other.

**Comment by Jalle19** [comment]
@ksooo epic

**Comment by ksooo** [comment]
Again merge conflicts. :-/ It's getting PITA, guys...

**Comment by ksooo** [comment]
@Jalle19 : It would be really very nice if you'd not merge to many PVR related stuff until this PR got merged... ;-)

**Comment by Jalle19** [comment]
@ksooo unfortunately that's a necessary evil, although I get what you're saying.

**Comment by opdenkamp** [comment]
just ignore the merge conflicts until we're ready to merge this in. you 
don't have to keep things rebased all the time.

**Comment by ksooo** [comment]
@opdenkamp any chance to get this merged this weekend, given that I rebase tonight?

**Comment by opdenkamp** [comment]
just told @jalle19 today that i'll be looking at PRs like this in the 
week end ;)

**Comment by ksooo** [comment]
Rebased.

**Comment by opdenkamp** [comment]
didn't get it checked out yesterday evening, as I spent most of the evening getting a properly working linux on my laptop. will do tonight

File: xbmc/pvr/timers/PVRTimers.cpp [hunk 1]
```diff
@@ -469,9 +469,9 @@ bool CPVRTimers::InstantTimer(const CPVRChannel &channel)
   if (!g_PVRManager.CheckParentalLock(channel))
     return false;
 
-  CEpgInfoTag epgTag;
-  bool bHasEpgNow = channel.GetEPGNow(epgTag);
-  CPVRTimerInfoTag *newTimer = bHasEpgNow ? CPVRTimerInfoTag::CreateFromEpg(epgTag) : NULL;
+  CEpgInfoTagPtr epgTag(channel.GetEPGNow());
+  bool bHasEpgNow = (epgTag != NULL);
```
Comment (opdenkamp) [inline xbmc/pvr/timers/PVRTimers.cpp]: this one won't be NULL (missing get() call)

**Comment by opdenkamp** [comment]
just one comment left that needs to be addressed, rest is fine. thanks for your patience :)

File: xbmc/pvr/timers/PVRTimers.cpp (hunk 1 above)
Comment (ksooo) [inline xbmc/pvr/timers/PVRTimers.cpp]: @opdenkamp Thought, that <code>bool operator!=(shared_ptr<T> const & p, std::nullptr_t);<code> would handle that correctly? Am I wrong?

Comment (Jalle19) [inline xbmc/pvr/timers/PVRTimers.cpp]: AFAIK you can compare shared pointers explicitly against null.

Comment (ksooo) [inline xbmc/pvr/timers/PVRTimers.cpp]: Googled a little. There seems not to be 100% agreement on that stuff like <code>bool bHasEpgNow = (epgTag != NULL);</code> works, but agreement is that <code>bool bHasEpgNow = (epgTag.get != NULL);</code> or <code>bool bHasEpgNow = epgTag;</code> (although the latter looks strange, it works due to shared_ptr's operator bool) will work. Thus, I will rescan the PR and change all respective code places accordingly.

**Comment by ksooo** [comment]
@opdenkamp: All requested changes are in now. Rebased and squashed down to a single commit. Ready for merging, I'd say.

**Comment by opdenkamp** [comment]
jenkins build and merge

**Comment by popcornmix** [comment]
Possibly crash caused by this:
http://forum.kodi.tv/showthread.php?tid=211501&pid=1924629#pid1924629

**Comment by AlwinEsch** [comment]
May be yes, has also one time a exception fault with my work on #6360, after a channel was deleted it crashed with this fault. In this time I thought comes from my work.

Has fixed it there with:
g_PVRChannelGroups->GetGroupAll(channel->IsRadio())->RemoveFromGroup(*channel);
https://github.com/AlwinEsch/kodi/blob/increase-pvr-addon-version/xbmc/pvr/dialogs/GUIDialogPVRChannelManager.cpp#L616

**Comment by opdenkamp** [comment]
can we get a trace from this with all symbols. i'm not seeing it here on my end

**Comment by ksooo** [comment]
cannot reproduce either.

**Pull request merged**